### PR TITLE
lxd/cgroups: enable cgroup2 limit support

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: lxd/cgroup/init.go [hunk 1]
```diff
@@ -6,20 +6,24 @@ import (
 	"path/filepath"
 	"strings"
 
+	lxc "gopkg.in/lxc/go-lxc.v2"
```
Comment (stgraber) [inline lxd/cgroup/init.go]: I want to keep this package free of `go-lxc`, can you instead have the check done in the ReadWriter we create in `container_lxc.go`?

Comment (brauner) [reply lxd/cgroup/init.go]: Oh right, I forgot about that constraint... Let me see...

File: lxd/container_lxc.go [hunk 2]
```diff
@@ -6966,12 +6985,17 @@ func (c *containerLXC) cgroup(cc *lxc.Container) (*cgroup.CGroup, error) {
 		rw.cc = c.c
 	}
 
-	return cgroup.New(&rw)
+	cg, err := cgroup.New(&rw)
+	if err == nil {
+		cg.UnifiedCapable = lxc.HasApiExtension("cgroup2")
+	}
+	return cg, err
```
Comment (stgraber) [inline lxd/container_lxc.go]: Would be cleaner as:
```go
cg, err := cgroup.New(&rw)
if err != nil {
    return nil, err
}

cg.UnifiedCapable = lxc.HasApiExtension("cgroup2")
return cg, nil
```

File: lxd/cgroup/init.go [hunk 3]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
...
+// controllers as well as new features for a given controller, i.e. you can
+// have "blkio" which is a controller and "blkio.weight" which is a feature of
+// the blkio controller.
+func (info *Info) Supports(resource Resources, cgroup *CGroup) bool {
+	val, ok := info.SupportsVersion(resource)
+	if val == V2 && cgroup != nil && !cgroup.UnifiedCapable {
+		ok = false
+	}
+	return ok
+}
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Comment says "knob" but code says "resource" and the type of "resource" (singular) is "Resources" (plural).
This seems a bit confusing :)

Comment (stgraber) [reply lxd/cgroup/init.go]: Types unless they're a slice, should always be singular.

File: lxd/cgroup/init.go [hunk 4]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
 	return "unknown"
 }
 
+// Resources is a generic type used to abstract resource control features
+// support for the legacy and unified hierarchy.
+type Resources int
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Ought to be `Resource` I think.

File: lxd/cgroup/init.go [hunk 5]
```diff
@@ -44,7 +45,6 @@ func GetInfo() Info {
 	info := Info{}
 	info.Namespacing = cgNamespace
 	info.Layout = cgLayout
-
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Why?

File: lxd/cgroup/init.go [hunk 6]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
 	return "unknown"
 }
 
+// Resources is a generic type used to abstract resource control features
+// support for the legacy and unified hierarchy.
+type Resources int
+
+const (
+	// Blkio resource control
+	Blkio Resources = iota
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Can you put in a line break after every const, will make parsing the associated comment a bit faster for my slow brain :)

File: lxd/cgroup/init.go [hunk 7]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
...
+// SupportsVersion indicates whether or not a given cgroup control knob is
+// available and in which type of cgroup filesystem.
+func (info *Info) SupportsVersion(resource Resources) (Backend, bool) {
+	switch resource {
+	case Blkio:
+		val, ok := cgControllers["blkio"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Line break before this and all subsequent similar returns please (to split the block which handles the retrieval+check from the map from the rest of the logic).

File: lxd/cgroup/init.go [hunk 8]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
...
+	case CPUSet:
+		val, ok := cgControllers["cpuset"]
+		if ok && val == V1 {
+			return V1, ok
+		}
+		return Unavailable, false
+	case Devices:
+		val, ok := cgControllers["devices"]
+		return val, ok
```
Comment (stgraber) [inline lxd/cgroup/init.go]: In such cases, I think `return cgControllers["devices"]` is actually cleaner

Comment (stgraber) [reply lxd/cgroup/init.go]: Assuming Go does the right thing in such cases, maps are a bit magic (can be one or two values dependent on context).

File: lxd/cgroup/init.go [hunk 9]
```diff
@@ -64,6 +64,204 @@ func (info *Info) Mode() string {
...
+// Note, we use "knob" instead of "controller" because this map holds
+// controllers as well as new features for a given controller, i.e. you can
+// have "blkio" which is a controller and "blkio.weight" which is a feature of
+// the blkio controller.
+func (info *Info) Supports(resource Resources, cgroup *CGroup) bool {
+	val, ok := info.SupportsVersion(resource)
+	if val == V2 && cgroup != nil && !cgroup.UnifiedCapable {
+		ok = false
+	}
+	return ok
```
Comment (stgraber) [inline lxd/cgroup/init.go]: Line break please

File: lxd/container_lxc.go [hunk 10]
```diff
@@ -1408,9 +1410,10 @@ func (c *containerLXC) deviceStaticShiftMounts(mounts []deviceConfig.MountEntryI
 
 // deviceAddCgroupRules live adds cgroup rules to a container.
 func (c *containerLXC) deviceAddCgroupRules(cgroups []deviceConfig.RunConfigItem) error {
+	cg, _ := c.cgroup(nil)
```
Comment (stgraber) [inline lxd/container_lxc.go]: We probably should still check `err` just to be safe, we don't want to change code in `cgroup` and suddenly be calling this function on `nil`.

**Comment by stgraber** [comment]
Looking good, mostly small nits.

File: lxd/cgroup/init.go (hunk 8 above)
Comment (brauner) [reply lxd/cgroup/init.go]: That doesn't work for maps...

Comment (brauner) [reply lxd/cgroup/init.go]:
```go
# github.com/lxc/lxd/lxd/cgroup
lxd/cgroup/init.go:159:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:161:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:163:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
lxd/cgroup/init.go:165:3: not enough arguments to return
        have (Backend)
        want (Backend, bool)
```

Comment (stgraber) [reply lxd/cgroup/init.go]: Yeah, was afraid of that, Go could do with a tiny bit more smarts around that case.

**Pull request merged**